### DISCUSSION TITLE
[API Proposal]: Add native support for exhaustive enum-based switch expressions

### DISCUSSION THREAD

**Discussion by skashtanoff** [dotnet/csharplang#7659] [Language Ideas]
### Background and motivation

Now (.NET 7), if we have a `enum`:

```csharp
enum State
{
    Off,
    On
}
```

Then when we use it in `switch`-expression:

```charp
var nextState = currentState switch
{
    State.Off => State.On,
    State.On  => State.Off 
};
```

The compiler will produce a warning/error despite of the fact the pattern is exhaustive.

### API Proposal

Any of the following would be acceptable

* allow the compiler to generate the default branch if the pattern is exhaustive. for example it could be `throw UnreachableException` by default
* add an attribute like `[ExhaustivePattern]` that turns the warning off. For example, we are sure that the value comes to us are `West` or `East`, and never equals to `North`, `South`, how about we tell about that to the compiler? maybe the JIT will manage to produce more optimal code based on this assumption

### API Usage

```csharp
// no warning here
var nextState = currentState switch
{
    State.Off => State.On,
    State.On  => State.Off 
};
```

### Risks

backward compatibility

**Comment by huoyaoyuan** [comment]
This is a C# language proposal, not API proposal. See https://github.com/dotnet/csharplang/issues/3179

**Comment by huoyaoyuan** [comment]
Some additions:

> * allow the compiler to generate the default branch if the pattern is exhaustive. for example it could be `throw UnreachableException` by default

The compiler already generates a branch throws `SwitchExpressionException`, with a warning about it.

> * add an attribute like `[ExhaustivePattern]` that turns the warning off.

Currently attributes can only be applied to constructs with compiled form in IL, like methods and fields. It can't be applied to statements inside a method.

**Comment by HaloFour** [comment]
See:  #3179
